source/agora/consensus/Ledger.d

```diff
@@ -1800,34 +1713,14 @@ unittest
     scope ledger = new TestLedger(genesis_validator_keys[0], blocks, params);
     ledger.simulatePreimages(Height(params.ValidatorCycle));
 
-    KeyPair kp = WK.Keys.A;
-    auto freeze_tx = genesisSpendable().front().refund(kp.address).sign(OutputType.Freeze);
-    assert(ledger.acceptTransaction(freeze_tx) is null);
-    ledger.forceCreateBlock(false);
-
-    Block block = ledger.lastBlock().clone();
-    assert(block.header.height == 1);
-    assert(block.header.validators.setCount < block.header.validators.count);
-    assert(!ledger.hasMajoritySignature(block.header.height));
-    assert(ledger.externalize(ConsensusData.init) == "Previous height does not have majority signature");
+    // if less than 4 out of 6 validators sign it should not externlize
+    iota(4).each!(signed =>
+        assert(ledger.externalize(ConsensusData.init, signed)
+            == "The majority of validators hasn't signed this block"));
 
-    auto validators = ledger.getValidators(block.header.height);
-    foreach (i; block.header.validators.setCount..validators.length)
-    {
-        block.header.validators[i] = true;
-        auto tmp = block.header.sign(WK.Keys[validators[i].address].secret, block.header.preimages[i]);
-        block.header.signature.R += tmp.R;
-        block.header.signature.s += tmp.s;
-
-        ledger.updateBlockMultiSig(block.header);
-
-        auto majority = block.header.validators.setCount > block.header.validators.count / 2;
-        assert(majority == ledger.hasMajoritySignature(block.header.height));
-    }
-
-    // should be able to create a new block after previous height receives majority signatures
-    ledger.forceCreateBlock();
-    assert(ledger.height() == 2);
+    // if at least 4 out of 6 validators sign it should externlize
+    iota(4, 7).each!(signed =>
+        assert(!ledger.externalize(ConsensusData.init, signed)));
```

Comment:
Don't we externalize 3 blocks in a row here ?

Reply:
Yes it would. Shall I add a ledger height assertion?

Reply:
Added the height assertions.